### ISSUE TITLE
conda-build mangles libcantera-devel build system files

### ISSUE THREAD

**Issue by ischoegl** [Cantera/conda-recipes#37]
PR #35 packages headers and libraries for both C++ and Fortran 90 in `libcantera-devel`. However, some build system files are currently deliberately excluded, as `conda-build` is unable to properly patch them.

One file where this is apparent is `lib/pgconfig/cantera.pc` (installed in `/opt/conda/envs/ct-env`), which contains references to the GH build system (in this case `/usr/share/miniconda3/conda-bld/cantera-recipe_1660271238971/_build_env`), i.e.

```
$ cat cantera.pc 
prefix=/opt/conda/envs/ct-env
exec_prefix=${prefix}/bin
libdir=${prefix}/lib
includedir=${prefix}/include

Name: Cantera
Description: Cantera library
URL: http://www.cantera.org
Version: 3.0.0a2

Libs: -L${libdir} -L/usr/share/miniconda3/conda-bld/cantera-recipe_1660271238971/_build_env/lib -lcantera -lpthread
Cflags: -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /opt/conda/envs/ct-env/include -pthread -I${includedir} -I/usr/share/miniconda3/conda-bld/cantera-recipe_1660271238971/_build_env/include -I/opt/conda/envs/ct-env/include
```

In this case, the original `cantera.pc.in` template is:

```
prefix=@pc_prefix@
exec_prefix=${prefix}/bin
libdir=${prefix}/@libdirname@
includedir=${prefix}/include

Name: Cantera
Description: Cantera library
URL: http://www.cantera.org
Version: @cantera_version@

Libs: -L${libdir} @pc_libdirs@ @pc_libs@
Cflags: @pc_cflags@ -I${includedir} @pc_incdirs@
```

indicating that the replacements `@pc_libdirs@` and `@pc_incdirs@` fail.

While `pkg-config` still works, other build system files are currently not included in the package (e.g. `Cantera.mak`).

There are two potential solutions:

1. Update template files upstream in `Cantera/cantera` in a way that all paths become patchable by `conda-build`
2. Strategically patch/replace those files within `Cantera/conda-recipes` before running `scons build` and `scons install`
3. Search/replace the mangled bits before packaging (e.g. using `sed`)

At least on *nix systems, (3) appears to be easiest?

**Comment by bryanwweber** [comment]
I suspect this is either because (1) Conda doesn't know to look at these files, or (2) Conda doesn't replace the build environment prefix, but instead the host environment prefix in the files. So these files should have the host environment in them, not the build environment. Why they have the latter and not the former is a mystery to me.

**Comment by ischoegl** [comment]
> So these files should have the host environment in them, not the build environment. Why they have the latter and not the former is a mystery to me.

Interesting. The files are widely used in testing, which may be part of the reason? If this is the case, then I revise my preferred resolution to (1) at the top, as - especially for samples whichever way they’re installed - the build environment should be moot.

**Comment by bryanwweber** [comment]
Yeah, I agree that's the preferred resolution, but I'm not sure if it's a problem with the files themselves, the recipe, or conda-build...

**Comment by ischoegl** [comment]
Fwiw, here is the `Makefile` packaged with a `Makefile` for the current `libcantera-devel` package (Cantera 2.6.0), which exhibits similar behavilr. This example is for `share/cantera/samples/cxx/demo/Makefile`:

```
$ cat Makefile
include /opt/conda/envs/ct-forge/include/cantera/Cantera.mak

CC=/home/conda/feedstock_root/build_artifacts/cantera-recipe_1652286095642/_build_env/bin/x86_64-conda-linux-gnu-cc
CXX=/home/conda/feedstock_root/build_artifacts/cantera-recipe_1652286095642/_build_env/bin/x86_64-conda-linux-gnu-c++
RM=rm -f
CXXFLAGS=-march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/ct-forge/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/cantera-recipe_1652286095642/work=/usr/local/src/conda/cantera-recipe-2.6.0 -fdebug-prefix-map=/opt/conda/envs/ct-forge=/usr/local/src/conda-prefix -pthread -O3 -Wno-inline -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /opt/conda/envs/ct-forge/include
CPPFLAGS=$(CANTERA_INCLUDES)
LDFLAGS=
LDLIBS=$(CANTERA_LIBS)

SRCS=demo.cpp
OBJS=$(subst .cpp,.o,$(SRCS))

all: demo

clean:
	$(RM) $(OBJS) demo

dist-clean: clean
	$(RM) *~ 
```

with the packaged `include/cantera/Cantera.mak` being:
<details>
  <summary>cat Cantera.mak</summary>

```
###############################################################################
# Include Snippet for Makefiles
#
#    To create Cantera C++ applications from the install environment
#    include this file into your Makefile environment by putting
#    the line "include Cantera.mak" in your Makefile.
#
#  Main Variables:
#
#      CANTERA_INCLUDES = Variable containing the include path
#      CANTERA_LIBS = List of libraries to include on the link line
#
#      CANTERA_FORTRAN_LIBS = list of libraries to link for Fortran programs
#      CANTERA_FORTRAN_MODS = Directory containing the F90 .mod files
#

CANTERA_VERSION=2.6.0

###############################################################################
#        CANTERA CORE
###############################################################################

# The directory where Cantera include files may be found.
#  Include files in application programs should start with:
#     #include "cantera/thermo.h"
#     #include "cantera/kernel/HMWSoln.h"

CANTERA_INSTALL_ROOT=/opt/conda/envs/ct-forge

CANTERA_CORE_INCLUDES=-I$(CANTERA_INSTALL_ROOT)/include

CANTERA_EXTRA_INCLUDES=-I/opt/conda/envs/ct-forge/include/eigen3 -I/home/conda/feedstock_root/build_artifacts/cantera-recipe_1652286095642/_build_env/include 

# Required Cantera libraries
CANTERA_CORE_LIBS=-pthread -L/opt/conda/envs/ct-forge/lib -lcantera -lsundials_cvodes -lsundials_ida -lsundials_nvecserial -lsundials_sunlinsollapackdense -lsundials_sunlinsollapackband -lfmt -lyaml-cpp -llapack -lblas -Wl,-rpath,/opt/conda/envs/ct-forge/lib

CANTERA_CORE_LIBS_DEP = /opt/conda/envs/ct-forge/lib/libcantera.a

CANTERA_EXTRA_LIBDIRS=-L/opt/conda/envs/ct-forge/lib -Wl,-rpath,/opt/conda/envs/ct-forge/lib -L/home/conda/feedstock_root/build_artifacts/cantera-recipe_1652286095642/_build_env/lib -Wl,-rpath,/home/conda/feedstock_root/build_artifacts/cantera-recipe_1652286095642/_build_env/lib

CANTERA_CORE_FTN=-L/opt/conda/envs/ct-forge/lib -lcantera_fortran -lcantera -lsundials_cvodes -lsundials_ida -lsundials_nvecserial -lsundials_sunlinsollapackdense -lsundials_sunlinsollapackband -lfmt -lyaml-cpp -llapack -lblas

CANTERA_FORTRAN_MODS=$(CANTERA_INSTALL_ROOT)/include/cantera

CANTERA_FORTRAN_SYSLIBS=-lpthread -lstdc++

###############################################################################
#            BOOST
###############################################################################

CANTERA_BOOST_INCLUDES=-I/opt/conda/envs/ct-forge/include

###############################################################################
#         CVODE/SUNDIALS LINKAGE
###############################################################################

CANTERA_SUNDIALS_INCLUDE=
CANTERA_SUNDIALS_LIBS= -lsundials_cvodes -lsundials_ida -lsundials_nvecserial -lsundials_sunlinsollapackdense -lsundials_sunlinsollapackband

###############################################################################
#         BLAS LAPACK LINKAGE
###############################################################################

CANTERA_BLAS_LAPACK_LIBS=-llapack -lblas

###############################################################################
#      COMBINATIONS OF INCLUDES AND LIBS
###############################################################################

CANTERA_INCLUDES=$(CANTERA_CORE_INCLUDES) $(CANTERA_SUNDIALS_INCLUDE) \
                 $(CANTERA_BOOST_INCLUDES) $(CANTERA_EXTRA_INCLUDES)

CANTERA_TOTAL_INCLUDES = $(CANTERA_INCLUDES)

# Add this into the compilation environment to identify the version number
CANTERA_DEFINES = -DCANTERA_VERSION=2.6.0

CANTERA_LIBS=$(CANTERA_CORE_LIBS) \
             $(CANTERA_EXTRA_LIBDIRS) $(CANTERA_SUNDIALS_LIBS) \
             $(CANTERA_BLAS_LAPACK_LIBS)

CANTERA_TOTAL_LIBS=$(CANTERA_LIBS)

CANTERA_TOTAL_LIBS_DEP= $(CANTERA_CORE_LIBS_DEP) \
                        $(CANTERA_SUNDIALS_LIBS_DEP)

CANTERA_FORTRAN_LIBS=$(CANTERA_CORE_FTN) \
                     $(CANTERA_EXTRA_LIBDIRS) $(CANTERA_SUNDIALS_LIBS) \
                     $(CANTERA_BLAS_LAPACK_LIBS) $(CANTERA_FORTRAN_SYSLIBS)

###############################################################################
#  END
###############################################################################
```
</details>

**Comment by ischoegl** [comment]
Linking discussion … https://github.com/conda/conda-build/issues/2677

**Comment by bryanwweber** [comment]
Of course, there's no resolution in that issue 🙄 this is maybe too much overhead, but I wonder if it makes sense to have a separate `cantera-examples` package, whose only purpose is to install the examples? I think in that case it would make sense to list the compiler in the host dependencies, which should resolve this problem...

**Comment by ischoegl** [comment]
There are some interesting ideas in that discussion, which may still be relevant.

Regarding #35, I think I'm happy to punt on `examples (I'm not sure we need a different package)`, i.e. not try to resolve things as the PR doesn't introduce a new problem. All of this is pre-existing ...

**Comment by bryanwweber** [comment]
Agreed, this does not need to affect #35. Good point about the make files for `libcantera-devel` headers.

**Comment by ischoegl** [comment]
The fix in Cantera/cantera#1366 works when I run `conda-build` locally using the recipe of #35 and add the SCons option `conda_build=y`.

**Comment by ischoegl** [comment]
Resolution still requires #35.
